### ISSUE TITLE
Derived class can unsafely separate `**kwargs` into individual arguments

### ISSUE THREAD

**Issue by bluerayi** [astral-sh/ty#3937]
### Summary

In a derived class, separating part of `*args` into individual arguments is considered a violation of the LSP.

The following override of the `say` method is considered not to violate the Liskov Substitution Principle.

```python
class Parent:
    def say(self, *args: int, **kwargs: int) -> None:
        print("Hello!")


class Child(Parent):
    def say(
        self,
        a: int = 0,
        b: int = 0,
        *args: int,
        **kwargs: int,
    ) -> None:
        print(f"My favorite number is {a + b}!")
```

The two arguments added in the derived class would be passed as `args` or `kwargs` when the method is called from the base class. The types of these arguments are the base types of the base class’s `args` and `kwargs` (or, to be more precise, they match exactly). Furthermore, since they have default values, the method can be called without providing any arguments. For these reasons, the code should be considered compatible with the base class’s method.

Nevertheless, `ty` rejects this code based on the Liskov Substitution Principle.

Of course, if I set the types of `args` and `kwargs` to `typing.Any`, no error will occur, but this does not mean that proper type checking is being performed—it simply ignores the issue.

As a side note, `mypy==2.1.0` accepts this code. Furthermore, if the types of `args` and `kwargs` are not `int` (for example, if they are `str`), it will detect this as an invalid override.

I used Python 3.14.6, `ty==0.0.56`. and `pyproject.toml` is:

```toml
# snip

[tool.ty.environment]
python-version = "3.14"

[tool.ty.terminal]
error-on-warning = true
```

### Version

ty 0.0.56 (3078f9fa2 2026-07-01)

**Comment by sharkdp** [comment]
I think that ty is actually right here, but unfortunately doesn't give any explanation. The parent class accepts a method call `.say(1, a=2)`, but the child class does not (*multiple values for argument 'a'*).

**Comment by bluerayi** [comment]
Ah, you're right. In the following case, no error occurs (in short, I should have just followed the coding convention that requires optional arguments to have names when defining functions).

```python
class Parent:
    def say(self, **kwargs: int) -> None:
        pass


class Child(Parent):
    def say(
        self,
        *,
        a: int = 0,
        b: int = 0,
        **kwargs: int,
    ) -> None:
        pass
```

However, this time the exact opposite problem occurred. Even with code like the following, ty allows it:

```python
class Parent:
    def say(self, **kwargs: int) -> None:
        pass


class Child(Parent):
    def say(
        self,
        *,
        a: bool = False,
        b: bool = False,
        **kwargs: int,
    ) -> None:
        pass
```

Under this definition, the call `x.say(a=42)` would be valid if `x: Parent`, but invalid if `x: Child`. However, the type checker does not catch this at the time of definition.

**Comment by sharkdp** [comment]
Hm, yes. That seems like a bug / missing feature, yes. I'm adding this as a child issue of https://github.com/astral-sh/ty/issues/166.

**Comment by charliermarsh** [comment]
I believe this was fixed by https://github.com/astral-sh/ruff/pull/27517.
